### PR TITLE
Fix issues with HTTP integration

### PR DESCRIPTION
**Pull request by lovenm**
Sorry that I didn't mention working on this in the issue before, I assumed (before reading the contributing guidelines) that since it still has the help-wanted tag that it's free to be worked on.

## What did you implement:

Closes https://github.com/serverless/serverless/issues/3840

## How did you implement it:

The AWS ApiGateway HTTP and HTTP_PROXY integrations did not have proper handling for querystrings/paths/headers in request parameters or headers/statusCodes in response yet, so I  implementated those.

## How can we verify it:

Create an event with the http or http-proxy integration in serverless.yml with the http or http_proxy integration type with the aforementioned parameters.
```
- http:
    integration: http
    path: example/{foo}
    method: get
    request:
      parameters:
        paths:
          foo: true
    response:
      headers:
        Foo: "'Bar'"
      statusCodes:
          400:
              pattern: '4\d{2}'
              headers: 
                Bar: "'Baz'"

```
## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/plugins/aws/package/compile/events/apiGateway/lib/method/integration.js [hunk 1]
```diff
@@ -88,6 +88,12 @@ module.exports = {
       });
     }
 
+    if (type === 'AWS' || type === 'HTTP') {
+      _.assign(integration, {
+        RequestParameters: this.getIntegrationRequestParameters(http),
```
Comment (HyperBrain) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/method/integration.js]: You set the RequestParameters property unconditionally here - so, even if I do not specify one (intentionally) it gets set. According to the AWS REST API documentation, the property should be omitted if it is not used.

Imo the property should only be set in `integration` if the user specified it.

Comment (lovenm) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/method/integration.js]: Yeah, I agree. An oversight on my part.

**Comment by lovenm** [comment]
I noticed a small hitch today that I had missed previously, with the HTTP_PROXY Integration Request parameters handling. It also requires the RequestParameters set in Integration for path parameters to work correctly and that was only done for the HTTP and AWS integration types. I'll push a commit to fix that in a minute.

**Comment by lovenm** [comment]
@pmuens 
Any info on the status of this?

**Comment by ascandroli** [comment]
This PR solves the _[Invalid mapping expression specified: paths]._ for me.
Here is the service definition I'm using:

```
  mockups:
    handler: src/get.main
    events:
      - http:
          path: /{proxy+}
          method: get
          cors: true
          integration: http-proxy
          request:
            uri: "https://s3-${opt:region, self:provider.region}.amazonaws.com/mockups.${opt:stage, self:provider.stage}.doesntmatter/{proxy}.json"
            parameters:
              paths:
                proxy: true
``` 

I also attached the full serverless.yml file.

thanks @lovenm 

[serverless.yml.txt](https://github.com/serverless/serverless/files/1312209/serverless.yml.txt)

**Comment by pmuens** [comment]
Great! Thanks for giving this a spin @ascandroli 👍. That's really helpful!

File: lib/plugins/aws/package/compile/events/apiGateway/lib/validate.js [hunk 2]
```diff
@@ -66,49 +66,37 @@ module.exports = {
...
+                http.integration === 'AWS_PROXY' ? ['parameters'] : ['parameters', 'uri'];
+
+              if (!_.isEmpty(_.difference(keys, allowedKeys))) {
                 const requestWarningMessage = [
-                  'Warning! You\'re using the LAMBDA-PROXY in combination with a request',
+                  `Warning! You're using the ${http.integration} in combination with a request`,
                   ` configuration in your function "${functionName}". Only the`,
-                  ' \'request.parameters\' configs are available in conjunction with',
-                  ' LAMBDA-PROXY. Serverless will remove this configuration automatically',
+                  _.map(allowedKeys, value => `request.${value}`),
```
Comment (brettjurgens) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/validate.js]: this winds up showing up as:
`Only therequest.parameters,request.uriconfigs are available in conjunction with HTTP_PROXY.`

can you add a space after `the`, before `configs`, and between the `allowedKeys` (maybe `.join(', ')`)?

Comment (lovenm) [reply lib/plugins/aws/package/compile/events/apiGateway/lib/validate.js]: Sure, I'll look into it a bit later today. Thanks!

**Comment by lovenm** [comment]
FWIW, here's the configuration I've been using in a work project:

```
fetchProps:
    handler: src/handler.fetchProps
    name: fetchProps-${self:provider.stage}
    events:
      - http:
          integration: http
          path: v1/{token}/properties.json
          method: get
          cors: true
          request:
            uri: https://${self:custom.${self:provider.stage}.propertiesBaseUrl}/{token}/properties.json
            parameters:
              paths:
                token: true
              querystrings:
                lang: false
          response:
            headers:
              Cache-Control: "'public, max-age=${self:custom.cacheTtlInSeconds}'"
              Content-Type: "'application/json;charset=UTF-8'"
```

**Comment by HyperBrain** [comment]
Hi @lovenm , sorry for the huuuuuggge delay. I will check the implementation soon - so far the code looks good, just needs some final testing.

@horike37 @RafalWilinski Can you check if this PR works as expected? As @pmuens stated above, it needs some deep testing.

**Comment by ozbillwang** [comment]
@brettjurgens 

This is really useful fix for http proxy integration, could we have some progresses on it and get it merged?

**Comment by horike37** [comment]
Thank you for pinging @lovenm 
I missed it and will do final review :+1:

**Pull request merged**